### PR TITLE
kvserver: rebalance ranges with one voter using joint configurations 

### LINKED ISSUE #108420: allocator: thrashing behavior when replication factor is set to 1
**Describe the problem**

When replication is set to one and the removal target is the leaseholder, the allocator rebalances by adding a new replica but we are not removing. And we are waiting for the range to be touched next time to perform lease transfer and remove the old replica. This leads to thrashing behavior and over-replication sometimes.

https://github.com/cockroachdb/cockroach/blob/9ed70271734d0871254626d7909cb48fcb348d78/pkg/kv/kvserver/allocator/plan/util.go#L54-L62

**To Reproduce**

What did you do? Describe in your own words.

https://github.com/cockroachdb/cockroach/blob/5e1c94962945edf9d1ce134c0f8e6e4340a5c896/pkg/kv/kvserver/asim/tests/testdata/rand/rand_ranges#L238C1-L272C1

Jira issue: CRDB-30459

### LINKED ISSUE #124171: kvserver: scatter with rf=1 doesn't randomize replica placement
**Describe the problem**

When an `AdminScatterRequest` is issued against a range which has a replication factor of 1, the range to be scattered will add an additional replica before completing. This leaves the range with 2 replicas, one on the initial store and another on a random store.

**To Reproduce**

Restore a database with RF=1 into a multi-node cluster.

A targeted reproduction using split, then scatter is shown below.

```bash
roachprod create local -n 5
roachprod start local --binary=./artifacts/cockroach
# Create a test table with 1k splits which we will use to split+scatter.
roachprod sql local:1 -- -e "CREATE TABLE t (i INT PRIMARY KEY);"
roachprod sql local:1 -- -e "INSERT INTO t select generate_series(1,1000);"
roachprod sql local:1 -- -e "ALTER TABLE t CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
sleep 10
# Stop the replicate/lease queue from interfering with the test. The range for
# the test table should be down-replicated to just node 1 by now.
roachprod sql local:1 -- -e "SELECT lease_holder, count(*) FROM [SHOW RANGES FROM TABLE t WITH DETAILS] GROUP BY lease_holder;"
roachprod sql local:1 -- -e "SET CLUSTER SETTING kv.replicate_queue.enabled = false"
roachprod sql local:1 -- -e "SET CLUSTER SETTING kv.lease_queue.enabled = false"
roachprod sql local:1 -- -e "ALTER TABLE t SPLIT AT SELECT i FROM t;"
roachprod sql local:1 -- -e "ALTER TABLE t CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "WITH ranges_info AS ( SHOW RANGES FROM TABLE t WITH DETAILS), store_replica_count AS ( SELECT unnest(replicas) AS store_id FROM ranges_info), store_lease_count AS ( SELECT lease_holder AS store_id FROM ranges_info), replica_counts AS ( SELECT store_id, COUNT(*) AS replica_count FROM store_replica_count GROUP BY store_id), lease_counts AS ( SELECT store_id, COUNT(*) AS lease_count FROM store_lease_count GROUP BY store_id), max_counts AS ( SELECT (SELECT MAX(replica_count) FROM replica_counts) AS max_replica_count, (SELECT MAX(lease_count) FROM lease_counts) AS max_lease_count) SELECT r.store_id, r.replica_count, repeat('#', CEIL(10.0 * r.replica_count / m.max_replica_count)::INT) AS replica_distribution, COALESCE(l.lease_count, 0) AS lease_count, repeat('#', CEIL(10.0 * COALESCE(l.lease_count, 0) / m.max_lease_count)::INT) AS lease_distribution FROM replica_counts r LEFT JOIN lease_counts l ON r.store_id = l.store_id CROSS JOIN max_counts m ORDER BY r.replica_count DESC;"
# This sleep isn't strictly necessary, but makes it obvious when the table was
# split and when it was later scattered when looking at timeseries.
sleep 10
# Remove the constraint so that the range can scatter unrestricted.
roachprod sql local:1 -- -e "ALTER TABLE t SCATTER;"
roachprod sql local:1 -- -e "WITH ranges_info AS ( SHOW RANGES FROM TABLE t WITH DETAILS), store_replica_count AS ( SELECT unnest(replicas) AS store_id FROM ranges_info), store_lease_count AS ( SELECT lease_holder AS store_id FROM ranges_info), replica_counts AS ( SELECT store_id, COUNT(*) AS replica_count FROM store_replica_count GROUP BY store_id), lease_counts AS ( SELECT store_id, COUNT(*) AS lease_count FROM store_lease_count GROUP BY store_id), max_counts AS ( SELECT (SELECT MAX(replica_count) FROM replica_counts) AS max_replica_count, (SELECT MAX(lease_count) FROM lease_counts) AS max_lease_count) SELECT r.store_id, r.replica_count, repeat('#', CEIL(10.0 * r.replica_count / m.max_replica_count)::INT) AS replica_distribution, COALESCE(l.lease_count, 0) AS lease_count, repeat('#', CEIL(10.0 * COALESCE(l.lease_count, 0) / m.max_lease_count)::INT) AS lease_distribution FROM replica_counts r LEFT JOIN lease_counts l ON r.store_id = l.store_id CROSS JOIN max_counts m ORDER BY r.replica_count DESC;"
```

This results in the following lease and replica distribution after scattering:

```
  store_id | replica_count | replica_distribution | lease_count | lease_distribution
-----------+---------------+----------------------+-------------+---------------------
         1 |          1001 | ##########           |         500 | ##########
         5 |           291 | ###                  |         147 | ###
         4 |           275 | ###                  |         137 | ###
         3 |           229 | ###                  |         118 | ###
         2 |           206 | ###                  |          99 | ##
```

Note how despite there being RF=1 and 1k ranges, there are 2k replicas, of which 1k remains on the original store (1). As a result of a replica always remaining on the replica where scatter was evaluated, that store also retains 1/2 the leases, leading to imbalanced lease counts which would affect restore processor planning.

**Expected behavior**

Scatter randomly places a replica and leaves the range with exactly 1 replica. 

**Environment:**
 - CockroachDB version: All versions affected

**Additional context**

See https://github.com/cockroachdb/cockroach/issues/108420 for explanation why rebalancing a voter with RF=1 will first add a voter.

Jira issue: CRDB-38742

### PR DESCRIPTION
**Pull request by kvoli**
The allocator would add a voter, instead of both adding and removing the
existing voter when rebalancing ranges with one replica. Removing the
leaseholder replica was not possible prior to https://github.com/cockroachdb/cockroach/pull/74077, so the addition
only was necessary.

This restriction is no longer necessary. Allow rebalancing a one voter
range between stores using joint configurations, where the lease will be
transferred to the incoming voter store, from the outgoing demoting
voter.

Scattering ranges with one voter will now leave the range with exactly
one voter, where previously both the leaseholder voter evaluating the
scatter, and the new voter would be left.

Before this patch, scattering 1000 ranges with RF=1 on a 5 store
cluster:

```
  store_id | replica_count | replica_distribution | lease_count | lease_distribution
-----------+---------------+----------------------+-------------+---------------------
         1 |          1001 | ##########           |         500 | ##########
         5 |           291 | ###                  |         147 | ###
         4 |           275 | ###                  |         137 | ###
         3 |           229 | ###                  |         118 | ###
         2 |           206 | ###                  |          99 | ##
```

After:

```
  store_id | replica_count | replica_distribution | lease_count | lease_distribution
-----------+---------------+----------------------+-------------+---------------------
         2 |           242 | ##########           |         241 | ##########
         4 |           227 | ##########           |         227 | ##########
         5 |           217 | #########            |         216 | #########
         3 |           209 | #########            |         208 | #########
         1 |           106 | #####                |         109 | #####
```

Fixes: https://github.com/cockroachdb/cockroach/issues/108420
Fixes: https://github.com/cockroachdb/cockroach/issues/124171

Release note (bug fix): Scattering a range with replication factor=1, no
longer erroneously up-replicates the range to two replicas. Leases will
also no longer thrash between nodes when perturbed with replication
factor=1.

### CODE REVIEW

**Comment by kvoli** [comment]
The existing leaseholder store is never selected as the rebalance target. That is to be expected initially, as scatter only jitters replica count stats, it doesn't completely randomize them. The leaseholder store has 1k+ replicas. However, after the replica counts get closer, we'd expect scatter to also do nothing 1/3 of the time but it is still continuing to choose other stores with this patch.

File: pkg/kv/kvserver/asim/tests/testdata/non_rand/example_add_node [hunk 1]
```diff
@@ -40,9 +40,9 @@ plot stat=replicas sample=1
  301 ┼──────────────────────────────────────╭────────────────────────────────────────
  281 ┤                                   ╭╭─╯
  261 ┤                                ╭╭──╯
- 241 ┤                              ╭╭─╯
- 221 ┤                           ╭───╯
- 201 ┤                        ╭╭─╯
+ 241 ┤                              ╭──╯
+ 221 ┤                           ╭──╯
+ 201 ┤                         ╭─╯╯
```
Comment (kvoli) [inline pkg/kv/kvserver/asim/tests/testdata/non_rand/example_add_node:45]: We see one rebalance slip in here, due to queue process timing. This is benign.

**Comment by kvoli** [comment]
> The existing leaseholder store is never selected as the rebalance target. That is to be expected initially, as scatter only jitters replica count stats, it doesn't completely randomize them. The leaseholder store has 1k+ replicas. However, after the replica counts get closer, we'd expect scatter to also do nothing 1/n of the time but it is still continuing to only choose other stores with this patch.

Turns out this is just due to timing. The allocator rebalance targets are returned before any rebalancing activity occurs. The disparity in replica counts initially causes the initial node to not be selected. If we iteratively ran scatter, the results would be closer to expectation w/ an even distribution. The lease/replicate queue will also adjust lease/replica counts as the scatter occurs.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r5, 6 of 7 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-:-NyMdQ0I35CMiI_RSLaQ:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter` line 0 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMb4vl9aoP42FclLhr:-NyMb4vl9aoP42FclLhs:bdeavht) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter#L0)):*
:fire: 
___
*[`pkg/kv/kvserver/allocator/plan/util_test.go` line 56 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMbTkb8JwCrCMYJ7Ku:-NyMbTkb8JwCrCMYJ7Kv:b319svz) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/kv/kvserver/allocator/plan/util_test.go#L56)):*
> ```Go
> 				{ChangeType: roachpb.REMOVE_VOTER, Target: roachpb.ReplicationTarget{NodeID: 1, StoreID: 1}},
> 			},
> 			expectedErrStr: "",
> ```

nit: `expectedPerformingSwap: false,`
___
*[`pkg/sql/scatter_test.go` line 215 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMcs2_CoiJi-PnYsFa:-NyMcs2_CoiJi-PnYsFb:bifs2qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/sql/scatter_test.go#L215)):*
> ```Go
> 		t, tc.ServerConn(0), "t",
> 		"k INT PRIMARY KEY, v INT",
> 		500,
> ```

nit: `/* numRows */`
___
*[`pkg/sql/scatter_test.go` line 262 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMdWLq3C25DF_WRnDx:-NyMdWLq3C25DF_WRnDy:b-hbglia) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/sql/scatter_test.go#L262)):*
> ```Go
> 	}
> 
> 	// Expect that the number of replicas on store 1 to have changed. We can't
> ```

It feels like we're leaving test flakiness up to chance here, but I guess `(1/3)^50` is a small enough number that we don't think it's at risk of flaking.

**Review by kvoli — COMMENTED** [review]
TYFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-:-NyMl6gNA4FfYekze5q_:bzftgo7)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter` line  at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMb4vl9aoP42FclLhr:-NyMl7sN0M3M_1B2djxa:b-awrxcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter#L)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:fire:
</blockquote></details>

Thanks :smile: 
___
*[`pkg/kv/kvserver/allocator/plan/util_test.go` line 56 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMbTkb8JwCrCMYJ7Ku:-NyMlBReDPw9C8HrhTj8:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/kv/kvserver/allocator/plan/util_test.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `expectedPerformingSwap: false,`
</blockquote></details>

Updated.
___
*[`pkg/sql/scatter_test.go` line 215 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMcs2_CoiJi-PnYsFa:-NyMlCDKAaR0vgA7tr6E:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/sql/scatter_test.go#L215)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `/* numRows */`
</blockquote></details>

Updated.
___
*[`pkg/sql/scatter_test.go` line 262 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/124284#-NyMdWLq3C25DF_WRnDx:-NyMlCwTDLB1TPtEHP8h:bm8owtw) ([raw file](https://github.com/cockroachdb/cockroach/blob/06635fc5e6a26f45d6f0ef3ef2010949ecf7fab1/pkg/sql/scatter_test.go#L262)):*
Seems reasonably small, in addition to:

> Turns out this is just due to timing. The allocator rebalance targets are returned before any rebalancing activity occurs. The disparity in replica counts initially causes the initial node to not be selected. If we iteratively ran scatter, the results would be closer to expectation w/ an even distribution. The lease/replicate queue will also adjust lease/replica counts as the scatter occurs.

**Comment by kvoli** [comment]
bors r=nvanbenschoten

**Pull request merged**